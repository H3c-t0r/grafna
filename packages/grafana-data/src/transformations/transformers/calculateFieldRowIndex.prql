# TODO: pick the table at the top of the pipeline
from A
# Row index
derive const = 0
derive `Row` = (row_number const) / (count const)
select {!const}
# Row percentile
derive const = 0
# TODO: Rename this if `Row` is already taken
derive `Row` = (row_number const) / (count const) #
select !{const}
