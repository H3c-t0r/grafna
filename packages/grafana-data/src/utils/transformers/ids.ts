export enum DataTransformerID {
  //  join = 'join', // Pick a field and merge all series based on that field
  append = 'append', // Merge all series together
  //  rotate = 'rotate', // Columns to rows
  reduce = 'reduce', // Run calculations on fields
  filter = 'filter', // Pick some fields
}
