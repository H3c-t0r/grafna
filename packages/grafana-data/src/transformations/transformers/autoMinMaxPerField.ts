import { DataTransformerID } from './ids';
import { DataTransformerInfo } from '../../types/transformations';
import { DataFrame, Field, FieldType, FieldConfig } from '../../types/dataFrame';
import { ReducerID, reduceField } from '../../transformations';
import isNumber from 'lodash/isNumber';

export interface AutoMinMaxPerFieldTransformerOptions {}

export const autoMinMaxPerFieldTransformer: DataTransformerInfo<AutoMinMaxPerFieldTransformerOptions> = {
  id: DataTransformerID.autoMinMaxPerField,
  name: 'Calculate min/max per field',
  description: `Calculate min/max for each field rather than for the complete set of data.
                Useful for cells in bar gauges.`,
  defaultOptions: {},
  transformer: (options: AutoMinMaxPerFieldTransformerOptions) => {
    const reducers = [ReducerID.min, ReducerID.max];
    return (data: DataFrame[]) => {
      console.log(data);
      return data.map((frame, _) => {
        const fields: Field[] = frame.fields.map(field => {
          if (field.type === FieldType.number && field.values) {
            console.log(field);
            console.log(field.config);
            const config: FieldConfig = { ...field.config };
            console.log(config);
            if (!isNumber(config.min) || !isNumber(config.max)) {
              const stats = reduceField({ field, reducers });
              config.min = stats[ReducerID.min];
              config.max = stats[ReducerID.max];
            }
            return {
              ...field,
              config,
            };
          } else {
            return field;
          }
        });
        return {
          ...frame,
          fields,
        };
      });
    };
  },
};
