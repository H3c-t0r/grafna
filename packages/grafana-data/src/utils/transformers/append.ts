import { DataTransformerInfo } from './transformers';
import { DataFrame } from '../../types/data';
import { DataTransformerID } from './ids';

interface AppendOptions {}

interface FieldIndex {
  [field: string]: number;
}

export const appendTransformer: DataTransformerInfo<AppendOptions> = {
  id: DataTransformerID.append,
  name: 'Append',
  description: 'Append all series',
  defaultOptions: {},

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: AppendOptions) => {
    return (data: DataFrame[]) => {
      if (data.length < 2) {
        return data;
      }

      const processed: DataFrame = {
        fields: [],
        rows: [],
      };
      const index: FieldIndex = {};
      let width = data[0].fields.length;
      for (let i = 0; i < width; i++) {
        const f = data[0].fields[i];
        processed.fields.push(f);
        index[f.name] = i;
      }

      for (const series of data) {
        const arr: number[] = [];
        for (let i = 0; i < series.fields.length; i++) {
          const f = series.fields[i];
          if (!index.hasOwnProperty(f.name)) {
            index[f.name] = width++;
            processed.fields.push(f);
          }
          arr[i] = index[f.name];
        }
        for (const row of series.rows) {
          const norm: any[] = [];
          for (let i = 0; i < row.length; i++) {
            norm[arr[i]] = row[i];
          }
          processed.rows.push(norm);
        }
      }

      return [processed];
    };
  },
};
