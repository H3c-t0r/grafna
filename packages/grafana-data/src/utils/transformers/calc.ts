import { DataTransformerInfo } from './transformers';
import { DataFrame, FieldType } from '../../types/data';
import { DataMatcherConfig, getDataMatcher } from '../matchers/matchers';
import { alwaysDataMatcher } from '../matchers/predicates';
import { DataTransformerID } from './ids';
import { ReducerID, fieldReducers, reduceField } from '../fieldReducer';

interface CalcOptions {
  calcs: string[];
  matcher?: DataMatcherConfig; // Assume all fields
}

export const calcTransformer: DataTransformerInfo<CalcOptions> = {
  id: DataTransformerID.calc,
  name: 'Calculate',
  description: 'calculate...',
  defaultOptions: {
    calcs: [ReducerID.min, ReducerID.max, ReducerID.mean, ReducerID.last],
  },

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: CalcOptions) => {
    const matcher = options.matcher ? getDataMatcher(options.matcher) : alwaysDataMatcher;
    const calculators = fieldReducers.list(options.calcs);
    const reducers = calculators.map(c => c.id);
    const fields = [
      {
        name: 'Field',
        type: FieldType.string,
      },
      ...calculators.map(info => {
        return {
          name: info.name,
        };
      }),
    ];

    return (data: DataFrame[]) => {
      const processed: DataFrame[] = [];
      for (const series of data) {
        if (matcher(series)) {
          const sub = {
            ...series,
            fields,
            rows: [] as any[], // empty rows
          };
          for (let i = 0; i < series.fields.length; i++) {
            const field = series.fields[i];
            if (matcher(series, field)) {
              const results = reduceField({
                series,
                fieldIndex: i,
                reducers,
              });
              const row: any[] = [];
              row.push(field.name);
              for (const s of reducers) {
                row.push(results[s]);
              }
              sub.rows.push(row);
            }
          }
          processed.push(sub);
        }
      }
      return processed;
    };
  },
};
