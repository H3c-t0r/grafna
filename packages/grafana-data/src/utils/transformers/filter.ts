import { DataTransformerInfo, NoopDataTransformer } from './transformers';
import { DataFrame, Field } from '../../types/dataFrame';
import { DataMatcherConfig, getDataMatcher } from '../matchers/matchers';
import { FieldMatcherID } from '../matchers/ids';
import { DataTransformerID } from './ids';

export interface FilterOptions {
  include?: DataMatcherConfig;
  exclude?: DataMatcherConfig;
}

export const filterTransformer: DataTransformerInfo<FilterOptions> = {
  id: DataTransformerID.filter,
  name: 'Filter',
  description: 'select a subset of fields',
  defaultOptions: {
    include: { id: FieldMatcherID.numericFields },
  },

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: FilterOptions) => {
    if (!options.include && !options.exclude) {
      return NoopDataTransformer;
    }

    const include = options.include ? getDataMatcher(options.include) : null;

    const exclude = options.exclude ? getDataMatcher(options.exclude) : null;

    return (data: DataFrame[]) => {
      const processed: DataFrame[] = [];
      for (const series of data) {
        // Find the matching field indexes
        const fields: Field[] = [];
        for (let i = 0; i < series.fields.length; i++) {
          const field = series.fields[i];
          if (exclude) {
            if (exclude(series, field)) {
              continue;
            }
            if (!include) {
              fields.push(field);
            }
          }
          if (include && include(series, field)) {
            fields.push(field);
          }
        }

        if (!fields.length) {
          continue;
        }
        const copy = {
          ...series, // all the other properties
          fields, // but a different set of fields
        };
        processed.push(copy);
      }
      return processed;
    };
  },
};
