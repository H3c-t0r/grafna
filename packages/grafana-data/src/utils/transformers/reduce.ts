import { DataTransformerInfo } from './transformers';
import { DataFrame, FieldType, Field } from '../../types/dataFrame';
import { DataMatcherConfig, getDataMatcher } from '../matchers/matchers';
import { alwaysDataMatcher } from '../matchers/predicates';
import { DataTransformerID } from './ids';
import { ReducerID, fieldReducers, reduceField } from '../fieldReducer';
import { KeyValue } from '../../types/data';
import { ArrayVector } from '../vector';
import { guessFieldTypeForField } from '../processDataFrame';

export interface ReduceOptions {
  reducers: string[];
  matcher?: DataMatcherConfig; // Assume all fields
}

export const reduceTransformer: DataTransformerInfo<ReduceOptions> = {
  id: DataTransformerID.reduce,
  name: 'Reducer',
  description: 'Return a DataFrame with the reduction results',
  defaultOptions: {
    calcs: [ReducerID.min, ReducerID.max, ReducerID.mean, ReducerID.last],
  },

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: ReduceOptions) => {
    const matcher = options.matcher ? getDataMatcher(options.matcher) : alwaysDataMatcher;
    const calculators = fieldReducers.list(options.reducers);
    const reducers = calculators.map(c => c.id);

    return (data: DataFrame[]) => {
      const processed: DataFrame[] = [];
      for (const series of data) {
        if (matcher(series)) {
          const values: ArrayVector[] = [];
          const fields: Field[] = [];
          const byId: KeyValue<ArrayVector> = {};
          values.push(new ArrayVector()); // The name
          fields.push({
            name: 'Field',
            type: FieldType.string,
            values: values[0],
            config: {},
          });
          for (const info of calculators) {
            const vals = new ArrayVector();
            byId[info.id] = vals;
            values.push(vals);
            fields.push({
              name: info.id,
              type: FieldType.other, // UNKNOWN until after we call the functions
              values: values[values.length - 1],
              config: {
                title: info.name,
              },
            });
          }
          for (let i = 0; i < series.fields.length; i++) {
            const field = series.fields[i];
            if (matcher(series, field)) {
              const results = reduceField({
                field,
                reducers,
              });
              // Update the name list
              values[0].buffer.push(field.name);
              for (const info of calculators) {
                const v = results[info.id];
                byId[info.id].buffer.push(v);
              }
            }
          }
          for (const f of fields) {
            const t = guessFieldTypeForField(f);
            if (t) {
              f.type = t;
            }
          }
          processed.push({
            ...series, // Same properties, different fields
            fields,
            length: values[0].length,
          });
        }
      }
      return processed;
    };
  },
};
