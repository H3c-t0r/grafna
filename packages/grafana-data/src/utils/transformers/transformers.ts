import { DataFrame } from '../../types/dataFrame';
import { Registry, RegistryItemWithOptions } from '../registry';

export type DataTransformer = (data: DataFrame[]) => DataFrame[];

export interface DataTransformerInfo<TOptions = any> extends RegistryItemWithOptions {
  transformer: (options: TOptions) => DataTransformer;
}

export interface DataTransformerConfig<TOptions = any> {
  id: string;
  options: TOptions;
}

// Transformer that does nothing
export const NoopDataTransformer = (data: DataFrame[]) => data;

/**
 * Apply configured transformations to the input data
 */
export function transformDataFrame(options: DataTransformerConfig[], data: DataFrame[]): DataFrame[] {
  let processed = data;
  for (const config of options) {
    const info = dataTransformers.get(config.id);
    const transformer = info.transformer(config.options);
    const after = transformer(processed);

    // Add a key to the metadata if the data changed
    if (after && after !== processed) {
      for (const series of after) {
        if (!series.meta) {
          series.meta = {};
        }
        if (!series.meta.transformations) {
          series.meta.transformations = [info.id];
        } else {
          series.meta.transformations = [...series.meta.transformations, info.id];
        }
      }
      processed = after;
    }
  }
  return processed;
}

// Initalize the Registry

import { filterTransformer } from './filter';
import { appendTransformer } from './append';
import { calcTransformer } from './calc';

export const dataTransformers = new Registry<DataTransformerInfo>(() => [
  filterTransformer,
  appendTransformer,
  calcTransformer,
]);
