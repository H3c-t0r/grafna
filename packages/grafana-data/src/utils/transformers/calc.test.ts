import { transformDataFrame } from './transformers';
import { ReducerID } from '../fieldReducer';
import { DataTransformerID } from './ids';

const seriesWithValues = {
  fields: [{ name: 'A' }, { name: 'B' }],
  rows: [
    [1, 2], // 1
    [2, 3], //
    [3, 4], //
    [4, 5], //
    [5, 6], //
    [6, 7], //
    [7, 8], //
  ],
};

describe('Calc Transformer', () => {
  it('filters by include', () => {
    const cfg = {
      id: DataTransformerID.calc,
      options: {
        calcs: [ReducerID.min, ReducerID.max, ReducerID.mean, ReducerID.delta],
      },
    };
    const filtered = transformDataFrame([cfg], [seriesWithValues])[0];
    expect(filtered.fields.length).toBe(5);
    expect(filtered).toMatchSnapshot();
  });
});
