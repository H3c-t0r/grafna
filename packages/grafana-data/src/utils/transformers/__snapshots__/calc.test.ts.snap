// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calc Transformer filters by include 1`] = `
Object {
  "fields": Array [
    Object {
      "config": Object {},
      "name": "Field",
      "type": "string",
      "values": Array [
        "A",
        "B",
      ],
    },
    Object {
      "config": Object {
        "title": "First",
      },
      "name": "first",
      "type": "string",
      "values": Array [
        "A",
        "B",
      ],
    },
    Object {
      "config": Object {
        "title": "Min",
      },
      "name": "min",
      "type": "string",
      "values": Array [
        "A",
        "B",
      ],
    },
    Object {
      "config": Object {
        "title": "Max",
      },
      "name": "max",
      "type": "string",
      "values": Array [
        "A",
        "B",
      ],
    },
    Object {
      "config": Object {
        "title": "Delta",
      },
      "name": "delta",
      "type": "string",
      "values": Array [
        "A",
        "B",
      ],
    },
  ],
  "labels": undefined,
  "meta": Object {
    "transformations": Array [
      "calc",
    ],
  },
  "name": undefined,
  "refId": undefined,
}
`;
